plutus-tx/src/PlutusTx/Ratio.hs

```diff
@@ -4,80 +4,349 @@
+    (n, d) -> if d P.== P.zero
+              then Builtins.error ()
+              else unsafeRatio n d
+
+-- | This mimics the behaviour of Aeson's instance for 'GHC.Rational'.
+instance ToJSON Rational where
+  toJSON (Rational n d) =
+    object
+      [ ("numerator", toJSON n)
+      , ("denominator", toJSON d)
+      ]
+
+-- | This mimics the behaviour of Aeson's instance for 'GHC.Rational'.
+instance FromJSON Rational where
+  parseJSON = withObject "Rational" Haskell.$ \obj -> do
+    n <- obj .: "numerator"
+    d <- obj .: "denominator"
+    case ratio n d of
+      Haskell.Nothing -> Haskell.fail "Zero denominator is invalid."
+      Haskell.Just r  -> Haskell.pure r
+
+-- | Makes a 'Rational' from a numerator and a denominator.
+--
+-- = Important note
+--
+-- If given a zero denominator, this function will error. If you don't mind a
+-- size increase, and care about safety, use 'ratio' instead.
+{-# INLINEABLE unsafeRatio #-}
+unsafeRatio :: Integer -> Integer -> Rational
+unsafeRatio n d
+  | d P.== P.zero = Builtins.error ()
+  | d P.< P.zero = unsafeRatio (P.negate n) (P.negate d)
+  | P.True =
+    let gcd' = euclid n d
+     in Rational (n `Builtins.quotientInteger` gcd')
+                 (d `Builtins.quotientInteger` gcd')
+
+-- | Safely constructs a 'Rational' from a numerator and a denominator. Returns
+-- 'Nothing' if given a zero denominator.
+{-# INLINEABLE ratio #-}
+ratio :: Integer -> Integer -> P.Maybe Rational
+ratio n d
+  | d P.== P.zero = P.Nothing
+  | d P.< P.zero = P.Just (unsafeRatio (P.negate n) (P.negate d))
+  | P.True =
+    let gcd' = euclid n d
+     in P.Just P..
+        Rational (n `Builtins.quotientInteger` gcd') P.$
+        d `Builtins.quotientInteger` gcd'
+
+-- | Converts a 'Rational' to a GHC 'Ratio.Rational', preserving value. Does not
+-- work on-chain.
+toGHC :: Rational -> Ratio.Rational
+toGHC (Rational n d) = n Ratio.% d
+
+-- | Returns the (possibly reduced) numerator of its argument.
+{-# INLINEABLE numerator #-}
+numerator :: Rational -> Integer
+numerator (Rational n _) = n
+
+-- | Returns the (possibly reduced) denominator of its argument. This will
+-- always be greater than 1, although the type does not describe this.
+{-# INLINEABLE denominator #-}
+denominator :: Rational -> Integer
+denominator (Rational _ d) = d
+
+-- | 0.5
+{-# INLINEABLE half #-}
+half :: Rational
+half = Rational 1 2
+
+-- | Converts an 'Integer' into the equivalent 'Rational'.
+{-# INLINEABLE fromInteger #-}
+fromInteger :: Integer -> Rational
+fromInteger num = Rational num P.one
+
+-- | Converts a GHC 'Ratio.Rational', preserving value. Does not work on-chain.
+fromGHC :: Ratio.Rational -> Rational
+fromGHC r = unsafeRatio (Ratio.numerator r) (Ratio.denominator r)
+
+-- | Produces the additive inverse of its argument.
+--
+-- = Note
+--
+-- This is specialized for 'Rational'; use this instead of the generic version
+-- of this function, as it is significantly smaller on-chain.
```

Comment:
There used to be a function with the same name in this module (now in `PlutusTx.Numeric`), with the following signature: ``(AdditiveGroup a, Ord a) => a -> a``. Given that `Rational` is both an `AdditiveGroup` and `Ord`, that function can potentially be used with it. However, this is quite bad for size:

```
    On-chain size comparison for abs:                 OK
      644B (~0.5KiB) vs 767B (~0.5KiB) 
      ~16% smaller
    On-chain size comparison for abs (overloaded):    OK
      69B vs 767B (~0.5KiB) 
      ~91% smaller
```

While it would be nice to enable both efficiency _and_ a generic implementation, it's rather tricky, as the (more size-efficient) `abs` in this PR's version of `PlutusTx.Ratio` relies a lot on internal invariants of `Rational`. 

I will add an explanatory note to this effect.